### ISSUE TITLE
[topic] Automated Feedback

### ISSUE THREAD

**Issue by kytrinyx** [exercism/exercism.io#2312]
Goal: Understand the underlying problems that the conversations are trying to address.

High level points:

* slow feedback cycle; too few people commenting
* people get stuck
* hard to discover interesting solutions
* not everyone knows how to look for problems in code
* out of date test suites are a distraction

Implemented so far:

* prototype of static analysis to provide a small, hand-crafted set of automated nitpicks

I will be closing the associated issues.

I intend to keep editing this issue as new, related discussions occur, trying to summarize the most interesting and important points.

As we decide to try specific things to address various problems, we can break out individual issues for each feature, keeping this as a more big picture discussion.

_UPDATE:_

I don't think that automated feedback is going to be a core piece of the exercism experience, but I do think that there could be automated ways of getting some small, nice-to-have features.

I don't intend to focus on automated feedback in the near future, because we have more important issues that are core to the user experience.

-------

**static analysis** #1602, #1607 

This discussion resulted in a series of iterations in  that eventually became [rikki-](http://exercism.io/rikki), a prototype which is currently providing some initial feedback on the first exercise in the ruby track.

rikki-'s role is to grab the results of static analysis, choose one thing, and then post a pre-written nitpick about that thing.

If this continues to work well, it could be expanded to provide feedback in other languages.

Goals:

- automated, helpful feedback to the user about code smells or complexity
- shorten the feedback loop after submitting
- metrics as conversation starters

Concerns:

- automatic metrics could be noisy
- seeing lots of feedback could be overwhelming
- it can feel very impersonal

**discussion forum** #2280, #2276

The goal was to help people get unstuck. It seems like this is trying to address the problem where too few people are providing feedback. It might also be about people being unsure where to get help when they're unable to complete an exercise.

A major concern with a discussion forum is that it could divert efforts away from nitpicking, which is the underlying problem that caused the suggestion to come up in the first place.

**automatic hints** #2276

Hints could be seen as pre-emptive nitpicks. People could submit hints for particular exercises to a central store, and when people ask for a hint we could deliver a randomised one.

This would likely take a lot of work to get right, and people might find that there simply are not enough hints, just as there are not enough nitpicks.

**test results along with solutions** #2177,  #1710, #1757, #23

There are two versions of this. One is to run the tests against people's solutions in some sort of CI environment, and another is to record the test results as people solve the problem, and then upload that to the servers.

The goal is to make it obvious at a glance to nitpickers whether or not the solution passes the tests.

- technically challenging to sandbox 20 different language environments
- output from all the various testing tools are different and would need to be normalized
- test suites change and a solution that passed at the time of submitting it might no longer pass

It seems like a very expensive / resource intensive technical solution to something that doesn't seem to be a huge problem. OK, so sometimes the tests don't pass. There's probably a lot to talk about in those cases.

**rank solutions using metrics** #1928

It might be useful/interesting to compare your own solution against others using various metrics.

- could lead to subtle competition and to trying to game the metrics
- focus on metrics might distract from, rather than add to, conversations

**using metrics to find dissimilar solutions** #1921

Using static analysis to pull up solutions that are very different from your own could be a great conversation starter, and very instructive. We wouldn't have to expose the actual metrics that this is based on, since the point is just to find interesting things to show you.

**track changes to the test suite** #1657

Problem: often when nitpicking older solutions people will claim that the solution doesn't pass the test. That is true, but it did at the time of submitting it, so it really doesn't matter.

In other words, the problem isn't that the code isn't passing, but that we shouldn't have to have this discussion. Sonia Keys came up with a suggestion that has been implemented in the Go track for a while now, and which is working well there. We may want to think about using the same tactic in other languages.

**Comment by pminten** [comment]
Regarding Rikki. I don't believe static (or dynamic) analysis tools will help very much here. Verification tools are great for pointing out that there's a problem, they generally suck at describing the problem in the terms of language a novice programmer thinks in. So at best a tool like that helps for very constrained problems.

Running tests against solutions: it's probably a huge resource hog, yes and you need to set up sandboxing and all that. What would be an improvement though is a way to easily download someone else's solution along with the test suite so that nitpickers can quickly check if a solution satisfies the test suite and run the benchmarks (if any).

**Comment by ErikSchierboom** [comment]
Perhaps there could be a way to mark some solutions as being "good" or "elegant." Perhaps this should be the function of the example solution each exercise has. You could then point people to those submissions to learn about how other people approached the problem.

**Comment by mtakac** [comment]
> Regarding Rikki. I don't believe static (or dynamic) analysis tools will help very much here. Verification tools are great for pointing out that there's a problem, they generally suck at describing the problem in the terms of language a novice programmer thinks in. So at best a tool like that helps for very constrained problems.

We can at least catch most common code smells and hint to the user what could be done to improve the code. We don't need to use messages provided by the verification tools  - I don't see any reason why we couldn't come up with our own, more user friendly tips.

**Comment by pminten** [comment]
Oh, it can certainly be done for very specific cases. It's just that it will be an enormous amount of work since understanding arbitrary code trees is difficult. To get something that works for multiple exercises you'll need to either implement a lot of checks for each separate exercise or limit yourself to general style feedback. And when you limit yourself to general style feedback you may just as well run a regular linter and return the results, possibly with some mangling to add hints.

**Comment by kytrinyx** [comment]
I talked to a graduate student at a university in London who is implementing what is essentially rikki- for Haskell. He found issues to address by doing statistical analysis on 10 years worth of first year programming student solutions, among other things.

I don't think we need a generic automated nitpicker, but something that could help identify some of the most egregious errors in the earliest exercises could help. Maybe only in the languages that beginners typically flock to.

**Comment by kytrinyx** [comment]
> Perhaps there could be a way to mark some solutions as being "good" or "elegant."

I wouldn't mind marking some solutions as good or elegant, as long as it's more than one solution. I really don't want people to think that there's only one good approach, since they end up nitpicking to that approach instead of critiquing what is actually there in the code.

Perhaps if we use label voting (see #1957 for a related discussion) or some other way to have people suggest things that they think are elegant, some of the people who are "trusted" on a track can look through the suggestions and actually mark them as good examples. These could then be explored.

Since a lot of people essentially take the same approach then I would feel bad about highlighting someone specific--perhaps we can anonymize the good examples saying that many different people submit good samples, here are a few different approaches.

**Comment by pminten** [comment]
Should we really have specific code examples? Wouldn't it be more educational if we compile a list of sensible approaches highlighting the important bits of each approach, preferably with a discussion of which approach takes which trade-off? That way people get a list of ideas they can use to solve an exercise in multiple ways.

**Comment by kytrinyx** [comment]
I much prefer the "sensible approaches" list over specific examples.

**Comment by ErikSchierboom** [comment]
Sensible approaches is a much better idea, I agree.

**Comment by LegalizeAdulthood** [comment]
My idea with "does it pass the tests" was that the user would upload test results with their solution.  Other people seem to have the idea that some automated cloud infrastructure would be applied to the user's solution.  That requires lots of infrastructure and duplicates the environment that is already on the user's machine.  My idea was simpler: I submit a solution *and* my test results along with it.  At a glance my submission has everything a reviewer needs to evaluate whether or not my code passed the tests as *I* had them at the time I got the problem.  This also avoids the problem of whether or not the test suite has changed since I submitted my solution.  Because running the tests and getting the results is all handled on the client side, it could be rolled out within a language track on a language-by-language basis and doesn't require every language track to implement the feature at the same time, or at all.  The only thing that's needed on the server side is some way of receiving the results and displaying them with my submission.

**Comment by kytrinyx** [comment]
Oh, yes, that's definitely a simpler approach, I'm updating the top post to reflect that. Thanks for elaborating, @LegalizeAdulthood.

**Comment by pminten** [comment]
The way I would expect this to work best is to have an `exercism test` subcommand that automatically runs the correct programs to test the code and stores the log in a standard location where `exercism submit` can automatically pick it up.

There are two challenges with this though:

1. It requires some kind of plugin mechanism for the CLI, though this could be as simple as having a `.test_command` file in the track dir.
2. While it works great for us command line junkies I expect the more IDE focused users to be less thrilled. Though I guess you could still run the tests as you normally do and just before submission run `exercism test` once to generate the log file.

**Comment by kytrinyx** [comment]
Another solution would be to have an `exercism record` or `exercism capture` command, and pipe the output of any command to it (per discussion in #1710).

**Comment by LegalizeAdulthood** [comment]
The original issue I opened had it all worked out for C++ in a way that was IDE friendly and automatic without requiring any extra intervention on behalf of the user.  It would go like this:

1. User builds their code in the IDE
2. The build runs the unit tests for human consumption.
3. The build runs the unit tests for machine consumption.
4. The build feeds machine readable output of the unit tests to the exercism client.
5. The client uploads the test data when the user submits.

I don't see any reason why this approach wouldn't work for other languages, but it's my impression that the other language tracks haven't supplied the user with any sort of build framework.  For C++ this simply wasn't feasible, so we have a build system for the C++ language track.  Having a build system in place lets you do additional automation of this sort of stuff.

Most unit test frameworks have some way of creating test result data as XML.  C++ uses Boost.Test which has this capability.  If you can send XML to the client, then the client just needs an XSLT transformation to normalize the test data into a consistent form for submitting to the web site.

**Comment by pminten** [comment]
Build systems are pretty uncommon in other tracks because they tend to add a layer of complexity that's unnecessary and often quite alien to the language culture. Using a build system in Go for example really isn't common since there's a standard to run tests (`go test`), and that's for a compiled language (though a relatively new one). Pretty all the interpreted languages like Elixir and Python don't need a build system at all, you just run the test file.

I really like the fact that for most tracks you can just run a test very easily without having to learn how to use a build system.

**Comment by LegalizeAdulthood** [comment]
I'm not a go developer, but yeah for things like Python or JavaScript you don't need a build system.  For Java and C++, I don't know what you'd do without it.  I guess you could use one of the web C++ IDEs for the simple exercises in exercism, but for real projects that doesn't work out.  At any rate, since I was the one asking for this and I had a way of working it out for C++, my suggestion is that we try it there to get some experience with it and then adapt/adjust for other language tracks.  In this particular case, having the build system already present makes this easier for C++ than for other language tracks.

**Comment by kytrinyx** [comment]
I'm going to close this in favor of the discussion here: https://github.com/exercism/discussions/issues/65
